plan/physical_plan_builder.go

```diff
@@ -571,6 +572,228 @@ func (p *Join) convert2PhysicalPlanRight(prop *requiredProperty, innerJoin bool)
+		innerChild  Plan
+		conds       []expression.Expression
+	)
+	if left {
+		outerSchema = p.children[0].Schema()
+		innerChild = p.children[1]
+	} else {
+		outerSchema = p.children[1].Schema()
+		innerChild = p.children[0]
+	}
+	corCols := make([]*expression.CorrelatedColumn, 0, outerSchema.Len())
+	for _, col := range outerSchema.Columns {
+		corCol := &expression.CorrelatedColumn{Column: *col, Data: new(types.Datum)}
+		corCol.Column.ResolveIndices(outerSchema)
+		corCols = append(corCols, corCol)
+	}
+	selection := &Selection{baseLogicalPlan: newBaseLogicalPlan(Sel, p.allocator)}
+	selection.self = selection
+	selection.initIDAndContext(p.ctx)
+	selection.SetSchema(innerChild.Schema().Clone())
+	selection.SetChildren(innerChild)
+	if left {
+		conds = make([]expression.Expression, 0, len(p.EqualConditions)+len(p.RightConditions)+len(p.OtherConditions))
+		for _, cond := range p.RightConditions {
+			newCond := convertCol2CorCol(cond, corCols, outerSchema)
+			conds = append(conds, newCond)
+		}
+	} else {
+		conds = make([]expression.Expression, 0, len(p.EqualConditions)+len(p.LeftConditions)+len(p.OtherConditions))
+		for _, cond := range p.LeftConditions {
+			newCond := convertCol2CorCol(cond, corCols, outerSchema)
+			conds = append(conds, newCond)
+		}
+	}
+	for _, cond := range p.EqualConditions {
+		newCond := convertCol2CorCol(cond, corCols, outerSchema)
+		conds = append(conds, newCond)
+	}
+	for _, cond := range p.OtherConditions {
+		newCond := convertCol2CorCol(cond, corCols, outerSchema)
+		conds = append(conds, newCond)
+	}
+	selection.Conditions = conds
+	_, selection.canControlScan = selection.makeScanController(true)
+	return selection, corCols
+}
+
+func (p *Join) convert2IndexNestedLoopJoinLeft(prop *requiredProperty, innerJoin bool) (*physicalPlanInfo, error) {
+	lChild := p.children[0].(LogicalPlan)
+	if _, ok := p.children[1].(*DataSource); !ok {
+		return nil, nil
+	}
+	allLeft := true
+	for _, col := range prop.props {
+		if !lChild.Schema().Contains(col.col) {
+			allLeft = false
+		}
+	}
+	lProp := prop
+	if !allLeft {
+		lProp = &requiredProperty{}
+	} else {
+		lProp = replaceColsInPropBySchema(prop, lChild.Schema())
+	}
+	var lInfo *physicalPlanInfo
+	var err error
+	if innerJoin {
+		lInfo, err = lChild.convert2PhysicalPlan(removeLimit(lProp))
+	} else {
+		lInfo, err = lChild.convert2PhysicalPlan(convertLimitOffsetToCount(lProp))
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if lInfo.p == nil {
+		return nil, nil
+	}
+	selection, corCols := p.buildSelectionWithConds(true)
+	if !selection.canControlScan {
+		return nil, nil
+	}
+	rInfo, _ := selection.makeScanController(false)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	join := &PhysicalHashJoin{
+		LeftConditions: p.LeftConditions,
+		// TODO: decide concurrency by data size.
+		Concurrency:   JoinConcurrency,
+		DefaultValues: p.DefaultValues,
+		SmallTable:    1,
+	}
+	join.tp = "NestedLoopJoinLeft"
```

Comment:
I found that many physical plan's tp is not a const, convert all of them to const?